autonomous/autonomous.py

```diff
@@ -1,10 +1,248 @@
+
+        # find goal waypoint index
+        # goal = self.goal + (not self.indexer.has_ball())
+        goal = self.goal + is_done  # for testing
+        if not goal == self.goal:  # if we want to move stop point
+            self.trap_profile, self.stop_point = self.generate_trap_profile(
+                goal, linear_state
+            )
+            print(
+                f"regenerating idx:{goal}, pos:{round(self.stop_point, 2)}, t: {round(tm, 3)}"
+            )
+            self.trap_profile_start_time = tm
+            # recalculate speed and position so we can use it in this loop
+            linear_state = self.trap_profile.calculate(0)
+            self.goal = goal
+
+        # find distance to start and end so look_around can be adjusted to not look beyond edges
+        to_start = linear_state.position
+        to_end = self.stop_point - linear_state.position
+        look_around = min(to_start, min(to_end, 0.5))
+        # find current goal pose
+        goal_pose = trajectory_generator.smooth_path(
+            self.waypoints, look_around, linear_state.position
+        )
+
+        cur_pose = self.chassis.odometry.getPose()
+        # the difference between last goal pose and this goal pose
+        goal_pose_diff = goal_pose.translation().distance(self.last_pose.translation())
+        # currentPose rotation and linearVelocityRef is only used for feedforward
+        self.chassis_speeds = self.drive_controller.calculate(
+            currentPose=cur_pose,
+            poseRef=goal_pose,
+            linearVelocityRef=goal_pose_diff,  # used for feedforward
+            angleRef=goal_pose.rotation(),
+        )
+        # send poses to driverstation
+        display_poses = [goal_pose, self.chassis.odometry.getPose()]
+        self.field.getRobotObject().setPoses(
+            [trajectory_generator.goal_to_field(p) for p in display_poses]
+        )
+        self.chassis.drive_local(
+            self.chassis_speeds.vx, self.chassis_speeds.vy, self.chassis_speeds.omega
+        )
+
+        wpilib.SmartDashboard.putNumber("auto_vel", float(linear_state.velocity))
+
+        self.last_pose = goal_pose
+
+    @state
+    def stopped(self):
+        """Finished moving but still want to be trying to fire,
+        needed for second ball at terminal"""
+        self.shooter_control.fire_input()
+
+
+# balls positions are described in https://docs.google.com/document/d/1K2iGdIX5vyCDEaJtaLdUiC-ihC9xyGYjrKFfLbvpusI/edit
+# note these are positions of the balls, not where you should go to pick them up
+# so actual auto waypoints should be chosen manually, only using these as referances
+blue_balls = [
+    (-3.299, 2.116),  # left
+    (-3.219, -2.176),  # middle
+    (-0.681, -3.826),  # right
+    (-7.156, -3.010),  # terminal
+    (-7.924, -1.559),  # cargo line
+]
+red_balls = [
+    (-2.249, 3.169),  # left
+    (-3.771, -0.935),  # middle
+    (-0.850, -3792),  # right
+]
+start_positions = [
```

Comment:
Any particular reason for this list? I don't see it being used as a list anywhere from a quick glance.